File: go/src/Vec3.go
package main

import "math"

type Vec3 struct {
	x float64
	y float64
	z float64
}

func (self Vec3) Clamp(min float64, max float64) (out Vec3) {
	out.x = math.Max(math.Min(self.x, max), min)
	out.y = math.Max(math.Min(self.y, max), min)
	out.z = math.Max(math.Min(self.z, max), min)
	return
}

func (self Vec3) ScalarSub(s float64) (out Vec3) {
	out.x = self.x - s
	out.y = self.y - s
	out.z = self.z - s
	return
}

func (self Vec3) VecSub(v Vec3) (out Vec3) {
	out.x = self.x - v.x
	out.y = self.y - v.y
	out.z = self.z - v.z
	return
}

func (self Vec3) ScalarAdd(s float64) (out Vec3) {
	out.x = self.x + s
	out.y = self.y + s
	out.z = self.z + s
	return
}

func (self Vec3) VecAdd(v Vec3) (out Vec3) {
	out.x = self.x + v.x
	out.y = self.y + v.y
	out.z = self.z + v.z
	return
}

func (self Vec3) ScalarDiv(s float64) (out Vec3) {
	out.x = self.x / s
	out.y = self.y / s
	out.z = self.z / s
	return
}

func (self Vec3) VecDiv(v Vec3) (out Vec3) {
	out.x = self.x / v.x
	out.y = self.y / v.y
	out.z = self.z / v.z
	return
}

func (a Vec3) ScalarMul(s float64) (out Vec3) {
	out.x = a.x * s
	out.y = a.y * s
	out.z = a.z * s
	return
}

func (self Vec3) VecMul(v Vec3) (out Vec3) {
	out.x = self.x * v.x
	out.y = self.y * v.y
	out.z = self.z * v.z
	return
}

func (self Vec3) Mag() (mag float64) {
	mag = math.Sqrt(
		math.Pow(self.x, 2) +
			math.Pow(self.y, 2) +
			math.Pow(self.z, 2))
	return
}

func (self Vec3) Norm() (normalized Vec3) {
	normalized = self.ScalarDiv(self.Mag())
	return
}

func (self Vec3) Dot(v Vec3) (dp float64) {
	dp = self.x*v.x +
		self.y*v.y +
		self.z*v.z
	return dp
}

File: rust/src/vec3.rs
use std::fmt;

#[derive(Copy, Clone)]
pub struct Vec3 {
    pub x: f64,
    pub y: f64,
    pub z: f64
}

impl fmt::Display for Vec3 {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Vec3{{{:4.2}, {:4.2}, {:4.2}}}", self.x, self.y, self.z)
    }
}

//  ADD
pub fn vec_add(a: Vec3, b: Vec3) -> Vec3 {
    Vec3{
        x: a.x + b.x,
        y: a.y + b.y,
        z: a.z + b.z}
}

pub fn scalar_add(a: Vec3, b: f64) -> Vec3 {
    Vec3{
        x: a.x + b,
        y: a.y + b,
        z: a.z + b}
}

//  SUB
pub fn vec_sub(a: Vec3, b: Vec3) -> Vec3 {
    Vec3{
        x: a.x - b.x,
        y: a.y - b.y,
        z: a.z - b.z}
}

pub fn scalar_sub(a: Vec3, b: f64) -> Vec3 {
    Vec3{
        x: a.x - b,
        y: a.y - b,
        z: a.z - b}
}

//  DIV
pub fn vec_div(a: Vec3, b: Vec3) -> Vec3 {
    Vec3{
        x: a.x / b.x,
        y: a.y / b.y,
        z: a.z / b.z}
}

pub fn scalar_div(a: Vec3, b: f64) -> Vec3 {
    Vec3{
        x: a.x / b,
        y: a.y / b,
        z: a.z / b}
}

//  MUL
pub fn vec_mul(a: Vec3, b: Vec3) -> Vec3 {
    Vec3{
        x: a.x * b.x,
        y: a.y * b.y,
        z: a.z * b.z}
}

pub fn scalar_mul(a: Vec3, b: f64) -> Vec3 {
    Vec3{
        x: a.x * b,
        y: a.y * b,
        z: a.z * b}
}

pub fn clamp(a: Vec3, min: f64, max: f64) -> Vec3 {
    Vec3{
        x: a.x.max(min).min(max),
        y: a.y.max(min).min(max),
        z: a.z.max(min).min(max)}
}

pub fn mag(a: Vec3) -> f64 {
    (a.x.powi(2) + a.y.powi(2) + a.z.powi(2)).sqrt()
}

pub fn norm(a: Vec3) -> Vec3 {
    let mag: f64 = mag(a);
    scalar_div(a, mag)
}

pub fn dot(a: Vec3, b: Vec3) -> f64 {
    a.x*b.x + a.y*b.y + a.z*b.z
}